File: 2-state/java/example-painter-v2/Canvas.java
public class Canvas {
    private Tool currentTool;

    public void setCurrentTool(Tool tool){
        this.currentTool = tool;
    }

    public void mouseDown(){
        this.currentTool.mouseDown();
    }
    public void mouseUp(){
        this.currentTool.mouseUp();
    }
    
}

File: 2-state/js/src/Canvas.js
const Tool = require('./Tool.js');

module.exports = class Canvas{
    #currentTool;

    changeTool(tool) {
        if (! tool instanceof Tool) throw new Error('Invalid tool');
        this.#currentTool = tool;
    }

    mouseDown() {
        this.#currentTool.mouseDown();
    }
    mouseUp() {
        this.#currentTool.mouseUp();
    }
}